### PR TITLE
[spy-plugin-registry-teardown] PR: Migrate application teardown to use the new teardown mechanism

### PR DESCRIPTION
**Pull request by andfoy**
Application plugin migration to use the new teardown mechanism

### Stacked PR Chain: [spy-plugin-registry-teardown]
| PR | Title |  Merges Into  |
|:--:|:------|:-------------:|
|#16012|[spy-plugin-registry-teardown] PR: Add plugin teardown operations to the plugin registry|**N/A**|
|#1|[spy-plugin-registry-teardown] PR: Migrate application teardown to use the new teardown mechanism|#16012|
|#2|[spy-plugin-registry-teardown] PR: Migrate breakpoints to use the new teardown mechanism|#1|
|#3|[spy-plugin-registry-teardown] PR: Migrate completions plugin to use the new teardown mechanism|#2|
|#4|[spy-plugin-registry-teardown] PR: Migrate Console, Explorer and Find in files to use the new teardown mechanism|#3|
|#5|[spy-plugin-registry-teardown] PR: Migrate help to use the new teardown mechanism|#4|
|#6|[spy-plugin-registry-teardown] PR: Migrate history and layouts to use the new teardown mechanism|#5|
|#7|[spy-plugin-registry-teardown] PR: Migrate Outline explorer, plots, preferences and profiler to use the new teardown mechanism|#6|
|#8|[spy-plugin-registry-teardown] PR: Migrate projects to use the new teardown mechanism|#7|
|#9|[spy-plugin-registry-teardown] PR: Migrate pylint, run and shortcuts to use the new teardown mechanism|#8|
|#10|[spy-plugin-registry-teardown] PR: Migrate statusbar, toolbar and tours to use the new teardown mechanism|#9|
|#11|[spy-plugin-registry-teardown] PR: Migrate variable explorer and working directory to use the new teardown mechanism|#10|
|#12|[spy-plugin-registry-teardown] PR: Add preference page to enable and disable plugins|#11|

### CODE REVIEW

File: .github/workflows/test-linux.yml [hunk 1]
```diff
@@ -16,10 +16,10 @@ on:
       - '**.sh'
 
   pull_request:
-    branches:
-      - master
-      - 5.*
-      - 4.x
+    # branches:
```
Comment (ccordoba12) [inline .github/workflows/test-linux.yml:19]: Why is this necessary?

File: spyder/api/widgets/menus.py [hunk 2]
```diff
@@ -168,6 +168,19 @@ def add_action(self: T,
...
+            action = self._actions_map.pop(item_id)
+            position = None
+
+            for i, (_, act) in enumerate(self._actions):
+                if act == action:
+                    position = i
+                    break
+
+            if position is not None:
+                self._actions.pop(position)
```
Comment (ccordoba12) [inline spyder/api/widgets/menus.py:182]: I think you need to mark the menu as dirty by setting `self._dirty = True` here, so it is re-rendered the next time a user clicks on it.

File: spyder/plugins/application/plugin.py [hunk 3]
```diff
@@ -101,6 +103,33 @@ def on_editor_available(self):
...
+        editor = self.get_plugin(Plugins.Editor)
+        self.get_container().sig_load_log_file.disconnect(editor.load)
+
+    @on_plugin_teardown(plugin=Plugins.Console)
+    def on_console_teardown(self):
+        if self.is_plugin_available(Plugins.MainMenu):
+            self.report_action.setVisible(False)
+
+    @on_plugin_teardown(plugin=Plugins.Shortcuts)
+    def on_shortcuts_teardown(self):
```
Comment (ccordoba12) [inline spyder/plugins/application/plugin.py]: Although this is the exact opposite of `on_shortcuts_available`, I think it's not correct.

I say it because deactivating Shortcuts shouldn't remove all other entries added by this plugin to the Help menu. Perhaps the only thing that should be done here is to remove the shortcuts action from the app context menu (see `get_application_context_menu`).

File: spyder/plugins/application/plugin.py [hunk 4]
```diff
@@ -206,6 +235,61 @@ def _populate_help_menu_about_section(self):
...
+            menu_id=ApplicationMenus.Help)
+
+    def _depopulate_file_menu(self):
+        mainmenu = self.get_plugin(Plugins.MainMenu)
+        mainmenu.remove_item_from_application_menu(
+            ApplicationActions.SpyderRestart,
+            menu_id=ApplicationMenus.File)
+        mainmenu.remove_item_from_application_menu(
+            ApplicationActions.SpyderRestartDebug,
+            menu_id=ApplicationMenus.File)
```
Comment (ccordoba12) [inline spyder/plugins/application/plugin.py]: Please use a `for` loop to remove these actions.

File: spyder/plugins/mainmenu/plugin.py [hunk 5]
```diff
@@ -264,6 +264,52 @@ def add_item_to_application_menu(self, item: ItemType,
...
+                        hasattr(action, 'action_id')):
+                    this_item_id = action.action_id
+                elif (isinstance(action, SpyderMenu) or
+                        hasattr(action, 'menu_id')):
+                    this_item_id = action.menu_id
+                if this_item_id is not None and this_item_id == item_id:
+                    position = i
+                    break
+            if position is not None:
+                actions.pop(position)
```
Comment (ccordoba12) [inline spyder/plugins/mainmenu/plugin.py:337]: How are these non-migrated menus re-rendered after you remove `item_id`?

File: .github/workflows/test-linux.yml (hunk 1 above)
Comment (andfoy) [reply .github/workflows/test-linux.yml:19]: To run the CIs on my fork

**Review by ccordoba12 — APPROVED** [review]
Looks good to me now, thanks @andfoy!
